### PR TITLE
Add channel create and delete endpoints

### PR DESCRIPTION
**Pull request by nesbitjd**
Thanks so much for looking at this for me!

### CODE REVIEW

File: router/middleware/session.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+	return dg
+}
+
+// EstablishSession establishes the session
+func EstablishSession() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		token, exists := os.LookupEnv("BOT_TOKEN")
+		if !exists {
+			fmt.Println("Bot token not set")
```
Comment (meinenec) [inline router/middleware/session.go]: We probably don't want to use `fmt` from within gin. Instead we should either return something to the user via JSON or use logrus to log an error.

File: router/middleware/session.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+			return
+		}
+
+		dg, err := discordgo.New("Bot " + token)
+		if err != nil {
+			fmt.Println("error creating Discord session,", err)
```
Comment (meinenec) [inline router/middleware/session.go]: We probably don't want to use `fmt` from within gin. Instead we should either return something to the user via JSON or use logrus to log an error.

File: api/channel.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+type Channel struct {
+	Name string
+	Type string
+}
+
+// CreateChannel creates a channel
+func CreateChannel(c *gin.Context) {
+	channel := new(Channel)
+	err := c.Bind(channel)
+	// Error message did not return in Insomnia
```
Comment (meinenec) [inline api/channel.go]: Is this still applicable?

File: api/channel.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		c.Error(err)
+		c.AbortWithStatusJSON(http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	dg := middleware.RetrieveSession(c)
+
+	cType := parseChannelType(channel.Type)
+	guildID := os.Getenv("GUILD_ID")
```
Comment (meinenec) [inline api/channel.go]: Could we check that `GUILD_ID` is set?

You might want to use lookupenv instead of getenv

File: api/channel.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	channel := c.Param("channel")
+
+	dg := middleware.RetrieveSession(c)
+
+	guildID := os.Getenv("GUILD_ID")
+
+	channels, err := dg.GuildChannels(guildID)
+	if err != nil {
+		return
```
Comment (meinenec) [inline api/channel.go:75]: could we add an error message?

File: api/channel.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// DeleteChannel deletes a channel
+func DeleteChannel(c *gin.Context) {
+
+	channel := c.Param("channel")
+
+	dg := middleware.RetrieveSession(c)
+
+	guildID := os.Getenv("GUILD_ID")
```
Comment (meinenec) [inline api/channel.go]: Could we check that `GUILD_ID` is set?

You might want to use lookupenv instead of getenv

File: api/channel.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+	channels, err := dg.GuildChannels(guildID)
+	if err != nil {
+		return
+	}
+
+	todelete := ""
+	for _, sample := range channels {
```
Comment (meinenec) [inline api/channel.go]: Can we change `sample` to something more meaningful?

File: api/channel.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		if strings.EqualFold(sample.Name, channel) {
+			todelete = sample.ID
+		}
+	}
+
+	_, err = dg.ChannelDelete(todelete)
+	if err != nil {
+		logrus.Error("Error deleting channel: %w", err)
```
Comment (meinenec) [inline api/channel.go]: Instead of just logging, we should return something to the user so they know what went wrong.

**Review by meinenec — CHANGES_REQUESTED** [review]
Can you also make sure that before pushing, you run:
- `go vet`
- `go mod tidy`
- ` gofmt`

File: router/middleware/session.go (hunk 2 above)
Comment (nesbitjd) [reply router/middleware/session.go]: Changed to use JSON to return error to the user

File: router/middleware/session.go (hunk 1 above)
Comment (nesbitjd) [reply router/middleware/session.go]: Changed to use JSON to return error to the user

File: api/channel.go (hunk 3 above)
Comment (nesbitjd) [reply api/channel.go]: The error message is now returning properly in Insomnia.

File: api/channel.go (hunk 4 above)
Comment (nesbitjd) [reply api/channel.go]: ```GUILD_ID``` is now checked to see if set using the bool in lookupenv.

File: api/channel.go (hunk 6 above)
Comment (nesbitjd) [reply api/channel.go]: ```GUILD_ID``` is now checked to see if set using the bool in lookupenv.

File: api/channel.go (hunk 5 above)
Comment (nesbitjd) [reply api/channel.go:75]: An error message has been added.

File: api/channel.go (hunk 7 above)
Comment (nesbitjd) [reply api/channel.go]: Changed to ```channelsElement``` to show that it is an element of the list of ```channels``` that we're ranging across.

File: api/channel.go (hunk 8 above)
Comment (nesbitjd) [reply api/channel.go]: Changed to return the error via JSON to the user.

File: api/channel.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if !exists {
+		retErr := fmt.Errorf("Guild ID not set")
+		c.Error(retErr)
+		c.AbortWithStatusJSON(http.StatusInternalServerError, retErr.Error())
+		return
+	}
+
+	_, err = dg.GuildChannelCreate(guildID, channel.Name, cType)
+	if err != nil {
+		return
```
Comment (meinenec) [inline api/channel.go:45]: We should return an error to the user in this case.

File: api/channel.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+			todelete = channelsElement.ID
+		}
+	}
+
+	_, err = dg.ChannelDelete(todelete)
+	if err != nil {
+		retErr := fmt.Errorf("Error deleting channel: %w", err)
+		c.Error(retErr)
+		c.AbortWithStatusJSON(http.StatusBadRequest, retErr.Error())
+
```
Comment (meinenec) [inline api/channel.go:90]: We should return here

**Review by meinenec — CHANGES_REQUESTED** [review]

File: api/channel.go (hunk 9 above)
Comment (nesbitjd) [reply api/channel.go:45]: Error creating channel is now reported to the user.

File: api/channel.go (hunk 10 above)
Comment (nesbitjd) [reply api/channel.go:90]: A return is added.

**Review by meinenec — APPROVED** [review]

**Pull request merged**